### DISCUSSION TITLE
Unhandled error in Jenkins pipeline

### DISCUSSION THREAD

**Discussion by bandantonio** [retypeapp/retype#21] [Q&A]
I'm receiving the following error when trying to use retype in my ci/cd system:

```sh
+ npm run build

> docs@0.0.1 build /home/jenkins/workspace/docs_master
> retype build

events.js:187
      throw er; // Unhandled 'error' event
      ^
```

package.json:

```json
{
  "name": "docs",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "serve": "retype watch",
    "build": "retype build"
  },
  "devDependencies": {
    "retypeapp": "^1.5.0"
  }
}

```

Jenkins:

```groovy
stage("Build") {
        docker.image("ci/nodejs:1.0").inside {
            sh "npm i"
            sh "npm run build"
        }
        publishHTML(target: [
            allowMissing: false,
            keepAll     : true,
            reportDir   : "${docsFolderName}",
            reportFiles : "index.html",
            reportName  : "Documentation"
        ])
    }
```

**Comment by fabriciomurta** [comment]
Hello, can you add `--verbose` to the retype commands? 

```
retype build --verbose
```

The error you are getting is not really giving any clue on what might have gone wrong.

**Reply by bandantonio** [reply]
It works in terminal, I see much more but it Jenkins it has no effect :( Output is the same

**Reply by geoffreymcgill** [reply]
Thanks for the update. We're going to investigate and see if we can reproduce the issue. 

Just to confirm, has Retype has been installed the machine running the CI process and running the command `retype --version` returns a value (likely `1.5.0`)?

**Reply by fabriciomurta** [reply]
You probably need to turn up verbosity in jenkins. Or, maybe [view actual logs](https://www.jenkins.io/doc/book/system-administration/viewing-logs/)?

**Reply by bandantonio** [reply]
Thanks for the quick reply.
Terminal: 1.5.0
Jenkins: the same error, no effect

**Reply by bandantonio** [reply]
Maybe this will help a bit:

```

> docs@0.0.1 ret:version /home/jenkins/workspace/tation_docs_master
> retype --version

events.js:187
      throw er; // Unhandled 'error' event
      ^

Error: spawn /home/jenkins/workspace/tation_docs_master/node_modules/retypeapp/platforms/linux-x64/retype ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:264:19)
    at onErrorNT (internal/child_process.js:456:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
Emitted 'error' event on ChildProcess instance at:
    at Process.ChildProcess._handle.onexit (internal/child_process.js:270:12)
    at onErrorNT (internal/child_process.js:456:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  errno: 'ENOENT',
  code: 'ENOENT',
  syscall: 'spawn /home/jenkins/workspace/tation_docs_master/node_modules/retypeapp/platforms/linux-x64/retype',
  path: '/home/jenkins/workspace/tation_docs_master/node_modules/retypeapp/platforms/linux-x64/retype',
  spawnargs: [ '--version' ]
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! docs@0.0.1 ret:version: `retype --version`
npm ERR! Exit status 1
```

**Reply by geoffreymcgill** [reply]
Is `retype` installed on the Jenkins container?

Maybe adding the `sh "npm install retypeapp --global"` config is required. I'm not super familiar with Jenkins, so maybe this is just a stab in dark. 

```
        docker.image("ci/nodejs:1.0").inside {
            sh "npm i"
            sh "npm run build"
            sh "npm install retypeapp --global"
        }
```

**Reply by fabriciomurta** [reply]
`ENOENT` is exactly "no such file or directory".

Per the error message it looks like jenkins wants to find retype at that location but can't find it. And per the path, it really looks like it was an `npm install` without `--global`.

You may remove retype from `devDependencies` in package.json as well (using geoffrey's extra line to install it as a global tool); installing it as a local module -and- global one would just spend extra time.

**Reply by fabriciomurta** [reply]
Just out of curiosity, I experimented with your exact package.json in an empty directory here; a `npm i` installed retype at the local folder and I could run it if I used the path similar to jenkins pipeline tried.

In my case, I used `node_modules/retypeapp/platforms/win-x64/retype --version` (for I am running windows OS), and it worked.

In your case, the file not being there, where the path makes sense, is still a mystery; my local install also included the `linux-x64/retype` binary, which suggests something prevented the pipeline from fully installing retype package. The standalone NPM package takes up to 243MB in disk space, could that be the reason your pipeline didn't allow the package files in?

In case jenkins comes with the `dotnet` CLI preinstalled, you can save a lot of required space and download if you remove the dependency from **package.json** and use `dotnet tool install --global retypeapp` instead.

**Reply by geoffreymcgill** [reply]
@bandantonio Were you able to get Retype working with Jenkins? Let us know if you require any further assistance and we will investigate.

**Reply by bandantonio** [reply]
Hi @geoffreymcgill 
Sorry for my super long reply - no, I wasn't able to get Retype working with Jenkins. I was in a hurry at that time (I was on a hackathon) and I didn't have much time to investigate the root cause but I'm 90% sure that I did something wrong.

I will try to investigate the installation process in Jenkins step-by-step a bit later and I will share my observations with you afterward.

**Comment by Genzer** [comment]
Hi @bandantonio,

I encountered the same problem and I had found a solution to this.

Basically, Retype is not a pure Node.JS package. Its core is implemented using .NET and the `retypeapp` NPM package is just a simple CLI wrapper to the binary (you can explore this by going to `node_modules/retypeapp/platforms/`.

Using a single `nodejs` Docker image alone won't work as it lacks system libraries (Retype binary is a dynamically-linked executable). I managed to analyzed and developed a `Dockerfile` which could use Retype on Docker. Please have a look at https://gist.github.com/Genzer/22c40a2e3d65e13c81903e3dd6874c42.

**Reply by fabriciomurta** [reply]
Thanks Genzer for taking your time to beat the mystery out of this bag! So in the end it was just some solibs that .NET require and the image didn't include! Even though we tried our best to include all dependencies in the package, we could only go as far as .NET goes.

For the record, the missing system libraries/packages were `libgssapi-krb5-2` and `libssl-dev`.
